### PR TITLE
Improve authorization flow.

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
NOTE $Subject

Related PR:
https://github.com/wso2/carbon-identity-framework/pull/4722

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java [hunk 1]
```diff
@@ -143,4 +145,34 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) thr
         }
         return authenticationResult;
     }
+
+    /**
+     * Check whether default credentials are being used for ClientAuthentication.
+     *
+     * @return true if default credentials are being used.
+     */
+    public boolean isUsingDefaultCredentials() {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java:154]: hasDefaultCredentialsUsed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java [hunk 2]
```diff
@@ -143,4 +145,34 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) thr
...
+
+        String hashedPasswordFromConfigFile = AuthConfigurationUtil.getInstance().getClientAuthenticationHash(defaultAppName);
+
+        if (hashedPasswordFromConfigFile != null) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java:158]: Use Stringutils

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java [hunk 3]
```diff
@@ -143,4 +145,34 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) thr
...
+                StringBuilder sb = new StringBuilder();
+                for (byte b : byteValue) {
+                    sb.append(Integer.toString((b & 0xff) + 0x100, 16).substring(1));
+                }
+
+                if (hashedPasswordFromConfigFile.equals(sb.toString())) {
+                    return true;
+                }
+            } catch (NoSuchAlgorithmException e) {
+                throw new RuntimeException(e);
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java:173]: Let's not break the flow on this. Let's put a warning.

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 4]
```diff
@@ -238,4 +256,30 @@ private boolean isClientEmpty(AuthenticationContext authenticationContext) {
...
+     * @param authHandlerName   Name of the auth Handler
+     * @param requestUri    Request URI
+     *
+     * @return true if endpoint is allowed to skip authorization.
+     */
+    private boolean isAuthorizationSkipAllowedEndpoint(String authHandlerName, String requestUri) {
+
+        String[] authorizationSkipAllowedEndpointConfig = AuthConfigurationUtil.getInstance().getSkipAuthorizationAllowedEndpoints()
+                .get(authHandlerName);
+        if (authorizationSkipAllowedEndpointConfig == null) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:272]: Check whether it is null or empty

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 5]
```diff
@@ -163,6 +170,17 @@ public void invoke(Request request, Response response) throws IOException, Servl
...
             }
+        /*
+         If no authenticated user or application is not found, and if it is a secured endpoint, check whether endpoint
+         is allowed to skip authorization for the engaged auth handler.
+         */
+        } else if ( authenticationContext != null && authenticationContext.getResourceConfig().isSecured()
+                && !isAuthorizationSkipAllowedEndpoint(authenticationContext.getProperty(
+                ENGAGED_AUTH_HANDLER).toString(), request.getRequestURI())) {
+
+            // If not allowed to skip authorization, 403-forbidden response will be received.
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:181]: Let's throw 401

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 6]
```diff
@@ -238,4 +256,30 @@ private boolean isClientEmpty(AuthenticationContext authenticationContext) {
...
+
+    /**
+     * Checks the endpoint is allowed to skip authorization for the given auth handler.
+     *
+     * @param authHandlerName   Name of the auth Handler
+     * @param requestUri    Request URI
+     *
+     * @return true if endpoint is allowed to skip authorization.
+     */
+    private boolean isAuthorizationSkipAllowedEndpoint(String authHandlerName, String requestUri) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:268]: isAuthorizationskipped

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java]: Addressed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java]: Addressed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java]: Addressed

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java]: Addressed

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java [hunk 7]
```diff
@@ -143,4 +145,35 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) thr
...
+     *
+     * @return true if default credentials are being used.
+     */
+    public boolean hasDefaultCredentialsUsed() {
+
+        String hashedPasswordFromConfigFile = AuthConfigurationUtil.getInstance().getClientAuthenticationHash(defaultAppName);
+
+        if (StringUtils.isNotBlank(hashedPasswordFromConfigFile)) {
+            try {
+                MessageDigest dgst = MessageDigest.getInstance(hashingFunction);
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java:160]: Remove the hash calculation and add hash as a constant

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 8]
```diff
@@ -338,4 +344,33 @@ public static boolean isAuthHeaderMatch(MessageContext messageContext, String au
...
+        OMElement skipAuthorizationConfig = IdentityConfigParser.getInstance().getConfigElement(
+                AUTHORIZATION_CONTROL_ELE);
+        if (skipAuthorizationConfig != null) {
+            Iterator<OMElement> configs = skipAuthorizationConfig.getChildrenWithName(
+                    new QName(SKIP_AUTHORIZATION_ELE));
+            if (configs != null) {
+                while (configs.hasNext()) {
+                    OMElement config = configs.next();
+                    String authHandlerName = config.getAttributeValue(new QName(AUTH_HANDLER_ELE));
+                    String[] allowedAuthHandlerNames = config.getAttributeValue(
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:362]: rename allowed endpoints

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java (hunk 5 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:181]: let's keep 403

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java (hunk 7 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java]: Addressed with commit 4b3365278d9d367e012446318ce545739817081d

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 8 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java]: Addressed with commit 4b3365278d9d367e012446318ce545739817081d

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 9]
```diff
@@ -338,4 +344,33 @@ public static boolean isAuthHeaderMatch(MessageContext messageContext, String au
...
+                    skipAuthorizationAllowedEndpoints.put(authHandlerName, allowedEndpoints);
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieve Map of endpoint that allowed to skip authorization against the auth handler.
+     */
+    public Map<String, String[]> getSkipAuthorizationAllowedEndpoints() {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:373]:
```suggestion
    public Map<String, String[]> getSkipAuthorizationAllowedEndpoints() {
    
```

Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:373]: Generate the method comment properly. please check other places as well.

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 10]
```diff
@@ -238,4 +256,30 @@ private boolean isClientEmpty(AuthenticationContext authenticationContext) {
...
+     * @param authHandlerName   Name of the auth Handler
+     * @param requestUri    Request URI
+     *
+     * @return true if endpoint is allowed to skip authorization.
+     */
+    private boolean isAuthorizationSkipped(String authHandlerName, String requestUri) {
+
+        String[] authorizationSkipAllowedEndpointConfig = AuthConfigurationUtil.getInstance()
+                .getSkipAuthorizationAllowedEndpoints().get(authHandlerName);
+        if (authorizationSkipAllowedEndpointConfig == null) {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:272]: check for empty as well

Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:272]: If the `authorizationSkipAllowedEndpointConfig` will be empty, if empty value is configured for a auth handler. we should not allow to skip any endpoint such scenario. isn't?

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java (hunk 9 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:375]: Addressed

**Review by ashensw — APPROVED** [review]

**Review by sachiniWettasinghe — APPROVED** [review]

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java [hunk 11]
```diff
@@ -49,6 +49,8 @@ public class ClientAuthenticationHandler extends AuthenticationHandler {
     private final String CLIENT_AUTH_HEADER = "Client";
     private final String hashingFunction = "SHA-256";
     private final String SERVICE_PROVIDER_KEY = "serviceProvider";
+    private final String defaultAppName = "dashboard";
+    private final String defaultHashedPassword = "66cd9688a2ae068244ea01e70f0e230f5623b7fa4cdecb65070a09ec06452262";
```
Comment (isharak) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/ClientAuthenticationHandler.java:53]: What is the purpose of this defaultHashedPassword ?

**Pull request merged**